delphi/translators/for2py/rectify.py

```diff
@@ -3668,84 +3669,123 @@ def indent(elem, level=0):
         if level and (not elem.tail or not elem.tail.strip()):
             elem.tail = i
 
-def buildNewAST(filename: str):
+def buildNewASTfromXMLString(xmlString: str) -> ET.Element:
     """
-        Using the list of cleaned AST, construct a new XML AST and write to a file
+        This function process OFP generated XML and generates
+        a rectified version by recursively calling the appropriate
+        functions.
     """
-    # A variable for keep a track of number of traverse time
     traverse = 1
 
-    # Read AST from the OFP generated XML file
-    ast = ET.parse(filename)
-    # Get a root of a tree
-    root = ast.getroot()
+    ofpAST = ET.XML(xmlString)
     XMLCreator = RectifyOFPXML()
     # A root of the new AST
-    newRoot = ET.Element(root.tag, root.attrib)
+    newRoot = ET.Element(ofpAST.tag, ofpAST.attrib)
     # First add the root to the new AST list
-    for child in root:
+    for child in ofpAST:
         # Handle only non-empty elementss
         if child.text:
             cur_elem = ET.SubElement(newRoot, child.tag, child.attrib)
             XMLCreator.parseXMLTree(child, cur_elem, newRoot, newRoot, traverse)
-
+    
+    # Indent and structure the tree properly
     tree = ET.ElementTree(newRoot)
     indent(newRoot)
-    
+
     # Checks if the rectified AST requires goto elimination,
     # if it does, it does a 2nd traverse to eliminate and
     # reconstruct the AST once more
     while (XMLCreator.need_goto_elimination):
+        oldRoot = newRoot
         traverse += 1
 
         XMLCreator.boundary_identifier(XMLCreator.goto_label_with_case)
 
-        root = tree.getroot()	
-        newRoot = ET.Element(root.tag, root.attrib)
-        for child in root:	
+        newRoot = ET.Element(oldRoot.tag, oldRoot.attrib)
+        for child in oldRoot:	
             if child.text:	
                 cur_elem = ET.SubElement(newRoot, child.tag, child.attrib)	
                 XMLCreator.parseXMLTree(child, cur_elem, newRoot, newRoot, traverse)	
-        tree = ET.ElementTree(newRoot)	
+        tree = ET.ElementTree(newRoot)
         indent(newRoot)
         if not XMLCreator.continue_elimination:
             XMLCreator.need_goto_elimination = False
```

Comment:
Would it be correct to replace this if-statement with:
```
XMLCreator.need_goto_elimination = XMLCreator.continue_elimination
```
?

Reply:
@skdebray I will address this in the goto_elimination branch. Thank you!